File: bot-api/dotnet/Robocode.TankRoyale.BotApi.Tests/src/BotInfoTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Extensions.Configuration;
using NUnit.Framework;
using Robocode.TankRoyale.BotApi.Util;

using static Robocode.TankRoyale.BotApi.BotInfo;
using static Robocode.TankRoyale.BotApi.Util.CountryCode;

namespace Robocode.TankRoyale.BotApi.Tests;

public class BotInfoTest
{
    static readonly string Name = "  TestBot  ";
    static readonly string Version = "  1.0  ";
    static readonly List<string> Authors = new List<string> { " Author1  ", " Author2 " };
    static readonly string Description = "  description ";
    static readonly string Homepage = " https://testbot.robocode.dev ";
    static readonly List<string> CountryCodes = new List<string> { " gb ", "  US " };
    static readonly List<string> GameTypes = new List<string> { " classic ", " melee ", " 1v1 " };
    static readonly string Platform = " .Net 6 ";
    static readonly string ProgrammingLang = " C# 10 ";
    static readonly InitialPosition InitialPosition = InitialPosition.FromString("  10, 20, 30  ");

    // -- Constructor: All parameters --

    [Test]
    public void Constructor_OK_AllFieldSetToValidCommonValues()
    {
        var botInfo = CreateBotInfo();

        Assert.That(botInfo.Name, Is.EqualTo(Name.Trim()));
        Assert.That(botInfo.Version, Is.EqualTo(Version.Trim()));
        Assert.That(botInfo.Authors, Is.EqualTo(Authors.ConvertAll(str => str.Trim())));
        Assert.That(botInfo.Description, Is.EqualTo(Description.Trim()));
        Assert.That(botInfo.Homepage, Is.EqualTo(Homepage.Trim()));
        Assert.That(botInfo.CountryCodes.ToList().ConvertAll(str => str.ToUpper()),
            Is.EqualTo(CountryCodes.ConvertAll(str => str.ToUpper()).ConvertAll(str => str.Trim())));
        Assert.That(botInfo.GameTypes, Is.EqualTo(GameTypes.ConvertAll(str => str.Trim())));
        Assert.That(botInfo.Platform, Is.EqualTo(Platform.Trim()));
        Assert.That(botInfo.ProgrammingLang, Is.EqualTo(ProgrammingLang.Trim()));
        Assert.That(botInfo.InitialPosition, Is.EqualTo(InitialPosition));
    }

    // -- Constructor: Name --

    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_Throws_NameIsNullEmptyOrBlank(string name)
    {
        var builder = PrefilledBuilder().SetName(name);
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower(), Is.EqualTo("name cannot be null, empty or blank"));
    }

    [Test]
    public void Constructor_OK_NameIsMaxLength()
    {
        var name = StringOfLength(MaxNameLength);
        var botInfo = PrefilledBuilder().SetName(name).Build();
        Assert.That(botInfo.Name, Is.EqualTo(name));
    }

    [Test]
    public void Constructor_Throws_NameLengthIsTooLong() {
        var builder = PrefilledBuilder().SetName(StringOfLength(MaxNameLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("name length exceeds"), Is.True);
    }
    
    // -- Constructor: Version --

    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_Throws_VersionIsNullEmptyOrBlank(string version)
    {
        var builder = PrefilledBuilder().SetVersion(version);
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower(), Is.EqualTo("version cannot be null, empty or blank"));
    }
    
    [Test]
    public void Constructor_OK_VersionIsMaxLength()
    {
        var version = StringOfLength(MaxVersionLength);
        var botInfo = PrefilledBuilder().SetVersion(version).Build();
        Assert.That(botInfo.Version, Is.EqualTo(version));
    }
    
    [Test]
    public void Constructor_Throws_VersionLengthIsTooLong() {
        var builder = PrefilledBuilder().SetVersion(StringOfLength(MaxVersionLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("version length exceeds"), Is.True);
    }

    // -- Constructor: Authors --

    [Test]
    [TestCaseSource(nameof(ListOfEmptyOrBlanks))]
    public void Constructor_Throws_AuthorsAreInvalid(List<string> authors)
    {
        var builder = PrefilledBuilder().SetAuthors(authors);
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower(), Is.EqualTo("authors cannot be null or empty or contain blanks"));
    }

    [Test]
    public void Constructor_OK_AuthorIsMaxLength()
    {
        var author = StringOfLength(MaxAuthorLength);
        var botInfo = PrefilledBuilder().SetAuthors(new List<string> { author }).Build();
        Assert.That(botInfo.Authors[0], Is.EqualTo(author));
    }
    
    [Test]
    public void Constructor_Throws_AuthorLengthIsTooLong() {
        var builder = PrefilledBuilder().AddAuthor(StringOfLength(MaxAuthorLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("author length exceeds"), Is.True);
    }
    
    [Test]
    public void Constructor_OK_MaxNumberOfAuthors() {
        var builder = PrefilledBuilder().SetAuthors(null);
        for (var i = 0; i < MaxNumberOfAuthors; i++) {
            builder.AddAuthor(Authors[0]);
        }
        Assert.That(builder.Build().Authors.Count, Is.EqualTo(MaxNumberOfAuthors));
    }

    [Test]
    public void Constructor_Throw_MoreThanMaxNumberOfAuthor()
    {
        var builder = PrefilledBuilder().SetAuthors(null);
        for (var i = 0; i < MaxNumberOfAuthors + 1; i++) {
            builder.AddAuthor(Authors[0]);
        }
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("number of authors exceeds the maximum"), Is.True);
    }

    // -- Constructor: Description --

    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_DescriptionIsNull_DescriptionIsNullOrEmptyOrBlank(string description)
    {
        var builder = PrefilledBuilder().SetDescription(description);
        var botInfo = builder.Build();
        Assert.That(botInfo.Description, Is.Null);
    }

    [Test]
    public void Constructor_OK_DescriptionIsMaxLength()
    {
        var description = StringOfLength(MaxDescriptionLength);
        var botInfo = PrefilledBuilder().SetDescription(description).Build();
        Assert.That(botInfo.Description, Is.EqualTo(description));
    }
    
    [Test]
    public void Constructor_Throws_DescriptionLengthIsTooLong() {
        var builder = PrefilledBuilder().SetDescription(StringOfLength(MaxDescriptionLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("description length exceeds"), Is.True);
    }
    
    // -- Constructor: Homepage --

    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_HomePageIsNull_HomepageIsNullOrEmptyOrBlank(string homepage)
    {
        var builder = PrefilledBuilder().SetHomepage(homepage);
        var botInfo = builder.Build();
        Assert.That(botInfo.Homepage, Is.Null);
    }

    [Test]
    public void Constructor_OK_HomepageIsMaxLength()
    {
        var homepage = StringOfLength(MaxHomepageLength);
        var botInfo = PrefilledBuilder().SetHomepage(homepage).Build();
        Assert.That(botInfo.Homepage, Is.EqualTo(homepage));
    }
    
    [Test]
    public void Constructor_Throws_HomepageLengthIsTooLong() {
        var builder = PrefilledBuilder().SetHomepage(StringOfLength(MaxHomepageLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("homepage length exceeds"), Is.True);
    }
    
    // -- Constructor: Country codes --
    
    [Test]
    [TestCaseSource(nameof(ListOfEmptyOrBlanks))]
    public void Constructor_WhenCountryCodesAreInvalid_ThenLocalCountryCodeIsUsed(List<string> countryCodes)
    {
        var builder = PrefilledBuilder().SetCountryCodes(countryCodes);
        var botInfo = builder.Build();
        Assert.That(botInfo.CountryCodes, Is.EqualTo(new List<string> { GetLocalCountryCode() }));
    }

    [Test]
    public void Constructor_OK_CountryCodeIsValid()
    {
        var botInfo = PrefilledBuilder().SetCountryCodes(new List<string> { "dk" }).Build();
        Assert.That(botInfo.CountryCodes[0].ToLower(), Is.EqualTo("dk"));
    }

    [Test]
    [TestCase("d")]
    [TestCase("dnk")]
    [TestCase("xx")]
    public void Constructor_DefaultCountryCode_CountryCodeIsInvalid(string countryCode)
    {
        var botInfo = PrefilledBuilder().SetCountryCodes(null).AddCountryCode(countryCode).Build();
        Assert.That(botInfo.CountryCodes[0].ToLower(), Is.EqualTo(GetLocalCountryCode().ToLower()));
    }
    
    [Test]
    public void constructor_OK_MaxNumberOfCountryCodes() {
        var builder = PrefilledBuilder().SetCountryCodes(null);
        for (var i = 0; i < MaxNumberOfCountryCodes; i++) {
            builder.AddCountryCode(CountryCodes[0]);
        }
        Assert.That(builder.Build().CountryCodes.Count, Is.EqualTo(MaxNumberOfCountryCodes));
    }

    [Test]
    public void constructor_Throw_MoreThanMaxNumberOfCountryCodes()
    {
        var builder = PrefilledBuilder().SetCountryCodes(null);
        for (var i = 0; i < MaxNumberOfCountryCodes + 1; i++) {
            builder.AddCountryCode(CountryCodes[0]);
        }
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("number of country codes exceeds the maximum"), Is.True);
    }
    
    // -- Constructor: Game types --
    
    [Test]
    [TestCaseSource(nameof(ListOfEmptyOrBlanks))]
    public void Constructor_Throws_GameTypesContainsEmptyOrBlanks(List<string> gameTypes)
    {
        var botInfo = PrefilledBuilder().SetGameTypes(gameTypes).Build();
        Assert.That(botInfo.GameTypes.Count, Is.Zero);
    }

    [Test]
    public void Constructor_OK_GameTypeIsMaxLength()
    {
        var gameType = StringOfLength(MaxGameTypeLength);
        var botInfo = PrefilledBuilder().SetGameTypes(new List<string> { gameType }).Build();
        Assert.That(botInfo.GameTypes.Contains(gameType), Is.True);
    }
    
    [Test]
    public void Constructor_Throws_GameTypeLengthIsTooLong() {
        var builder = PrefilledBuilder().AddGameType(StringOfLength(MaxGameTypeLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("game type length exceeds"), Is.True);
    }
    
    [Test]
    public void Constructor_OK_MaxNumberOfGameTypes() {
        var builder = PrefilledBuilder().SetCountryCodes(null);
        for (var i = 0; i < MaxNumberOfCountryCodes; i++) {
            builder.AddCountryCode(CountryCodes[0]);
        }
        Assert.That(builder.Build().CountryCodes.Count, Is.EqualTo(MaxNumberOfCountryCodes));
    }

    [Test]
    public void Constructor_Throw_MoreThanMaxNumberOfGameTypes()
    {
        var builder = PrefilledBuilder().SetCountryCodes(null);
        for (var i = 0; i < MaxNumberOfCountryCodes + 1; i++) {
            builder.AddCountryCode(CountryCodes[0]);
        }
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("number of country codes exceeds the maximum"), Is.True);
    }

    // -- Constructor: Platform --
    
    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_DefaultPlatform_PlatformIsNullOrEmptyOrBlank(string platform)
    {
        var botInfo = PrefilledBuilder().SetPlatform(platform).Build();
        Assert.That(botInfo.Platform, Is.EqualTo(PlatformUtil.GetPlatformName()));
    }

    [Test]
    public void Constructor_OK_PlatformIsMaxLength() {
        var platform = StringOfLength(MaxPlatformLength);
        var botInfo = PrefilledBuilder().SetPlatform(platform).Build();
        Assert.That(botInfo.Platform, Is.EqualTo(platform));
    }

    [Test]
    public void Constructor_Throws_PlatformLengthIsTooLong() {
        var builder = PrefilledBuilder().SetPlatform(StringOfLength(MaxPlatformLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("platform length exceeds"), Is.True);
    }
    
    // -- Constructor: Programming language --
    
    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void Constructor_ProgrammingLangIsNull_ProgrammingLangIsNullOrEmptyOrBlank(string platform)
    {
        var builder = PrefilledBuilder().SetProgrammingLang(platform);
        var botInfo = builder.Build();
        Assert.That(botInfo.ProgrammingLang, Is.Null);
    }

    [Test]
    public void Constructor_OK_ProgrammingLangIsMaxLength() {
        var programmingLang = StringOfLength(MaxProgrammingLangLength);
        var botInfo = PrefilledBuilder().SetProgrammingLang(programmingLang).Build();
        Assert.That(botInfo.ProgrammingLang, Is.EqualTo(programmingLang));
    }

    [Test]
    public void Constructor_Throws_ProgrammingLangIsTooLong() {
        var builder = PrefilledBuilder().SetProgrammingLang(StringOfLength(MaxProgrammingLangLength + 1));
        var exception = Assert.Throws<ArgumentException>(() => builder.Build());
        Assert.That(exception?.Message.ToLower().Contains("programminglang length exceeds"), Is.True);
    }
    
    // -- Constructor: Initial position --
    
    [Test]
    [TestCase(null)]
    [TestCase("")]
    [TestCase("  ")]
    [TestCase("\t ")]
    [TestCase("\n")]
    public void constructor_InitialPositionIsNull_InitialPositionIsNullOrEmptyOrBlank(string initialPosition)
    {
        var builder = PrefilledBuilder().SetInitialPosition(InitialPosition.FromString(initialPosition));
        var botInfo = builder.Build();
        Assert.That(botInfo.InitialPosition, Is.Null);
    }

    // -- FromFile(filename) --

    [Test]
    public void FromFile_OK_ValidFilePath()
    {
        var filePath = Path.Combine(TestContext.CurrentContext.WorkDirectory, "../../../resources/TestBot.json");
        var botInfo = FromFile(filePath);
        Assert.That(botInfo.Name, Is.EqualTo("TestBot"));
        Assert.That(botInfo.Version, Is.EqualTo("1.0"));
    }

    [Test]
    public void FromFile_OK_ValidFileAndBasePath()
    {
        var basePath = Path.Combine(TestContext.CurrentContext.WorkDirectory, "../../../resources");
        var botInfo = FromFile("TestBot.json", basePath);
        Assert.That(botInfo.Name, Is.EqualTo("TestBot"));
        Assert.That(botInfo.Version, Is.EqualTo("1.0"));
    }

    [Test]
    public void FromFile_Throw_NonExistingFile()
    {
        const string filename = "non-existing-filename";
        Assert.Throws<FileNotFoundException>(() => FromFile(filename));
    }
    
    // -- FromConfiguration(filename) --

    [Test]
    public void FromConfiguration_OK_ValidFilePath()
    {
        var configBuilder = new ConfigurationBuilder()
            .SetBasePath(Path.Combine(TestContext.CurrentContext.WorkDirectory, "../../../resources"))
            .AddJsonFile("TestBot.json");

        var botInfo = FromConfiguration(configBuilder.Build());
        Assert.That(botInfo.Name, Is.EqualTo("TestBot"));
        Assert.That(botInfo.Version, Is.EqualTo("1.0"));
    }

    [Test]
    public void FromConfiguration_Throw_InvalidConfiguration()
    {
        Assert.That(() => FromConfiguration(new ConfigurationBuilder().Build()), Throws.Exception);
    }


    private static readonly object[] ListOfEmptyOrBlanks =
    {
        new object[] { new List<string>() },
        new object[] { new List<string> { "" } },
        new object[] { new List<string> { "\t" } },
        new object[] { new List<string> { " \n" } },
        new object[] { new List<string> { " ", "" } }
    };

    private static IBuilder PrefilledBuilder()
    {
        return Builder().Copy(new BotInfo(Name, Version, Authors, Description, Homepage, CountryCodes,
            GameTypes, Platform, ProgrammingLang, InitialPosition));
    }

    private static string StringOfLength(int length) => new string('x', length);
    
    private static BotInfo CreateBotInfo()
    {
        return new BotInfo(Name, Version, Authors, Description, Homepage, CountryCodes, GameTypes, Platform,
            ProgrammingLang, InitialPosition);
    }
}

File: bot-api/typescript/test/BotInfo.test.ts
import { describe, it, expect } from "vitest";
import { BotInfo } from "../src/BotInfo.js";

const NAME = "MyBot";
const VERSION = "1.0";
const AUTHORS = ["John Doe"];

describe("LEGACY", () => {
describe("BotInfo constants", () => {
  it("has correct max lengths", () => {
    expect(BotInfo.MAX_NAME_LENGTH).toBe(30);
    expect(BotInfo.MAX_VERSION_LENGTH).toBe(20);
    expect(BotInfo.MAX_AUTHOR_LENGTH).toBe(50);
    expect(BotInfo.MAX_DESCRIPTION_LENGTH).toBe(250);
    expect(BotInfo.MAX_HOMEPAGE_LENGTH).toBe(150);
    expect(BotInfo.MAX_GAME_TYPE_LENGTH).toBe(20);
    expect(BotInfo.MAX_PLATFORM_LENGTH).toBe(30);
    expect(BotInfo.MAX_PROGRAMMING_LANG_LENGTH).toBe(30);
    expect(BotInfo.MAX_NUMBER_OF_AUTHORS).toBe(20);
    expect(BotInfo.MAX_NUMBER_OF_COUNTRY_CODES).toBe(20);
    expect(BotInfo.MAX_NUMBER_OF_GAME_TYPES).toBe(10);
  });
});

describe("BotInfo construction", () => {
  it("creates with required fields", () => {
    const bot = new BotInfo(NAME, VERSION, AUTHORS, null, null, [], [], null, null, null);
    expect(bot.name).toBe(NAME);
    expect(bot.version).toBe(VERSION);
    expect(bot.authors).toEqual(AUTHORS);
  });

  it("builder pattern works", () => {
    const bot = BotInfo.builder(NAME, VERSION, AUTHORS)
      .setDescription("A test bot")
      .setHomepage("https://example.com")
      .setCountryCodes(["US"])
      .setGameTypes(["classic"])
      .setPlatform("Node.js")
      .setProgrammingLang("TypeScript")
      .build();
    expect(bot.name).toBe(NAME);
    expect(bot.description).toBe("A test bot");
    expect(bot.homepage).toBe("https://example.com");
    expect(bot.countryCodes).toEqual(["US"]);
    expect(bot.gameTypes).toEqual(["classic"]);
    expect(bot.platform).toBe("Node.js");
    expect(bot.programmingLang).toBe("TypeScript");
  });
});

describe("BotInfo validation", () => {
  it("stores null for blank name (convention-over-config: validated at connection time)", () => {
    const bot1 = new BotInfo("", VERSION, AUTHORS, null, null, [], [], null, null, null);
    expect(bot1.name).toBeNull();
    const bot2 = new BotInfo("   ", VERSION, AUTHORS, null, null, [], [], null, null, null);
    expect(bot2.name).toBeNull();
  });

  it("throws if name exceeds max length", () => {
    const longName = "a".repeat(31);
    expect(() => new BotInfo(longName, VERSION, AUTHORS, null, null, [], [], null, null, null)).toThrow("'name' length");
  });

  it("stores null for blank version (convention-over-config: validated at connection time)", () => {
    const bot = new BotInfo(NAME, "", AUTHORS, null, null, [], [], null, null, null);
    expect(bot.version).toBeNull();
  });

  it("throws if version exceeds max length", () => {
    const longVersion = "a".repeat(21);
    expect(() => new BotInfo(NAME, longVersion, AUTHORS, null, null, [], [], null, null, null)).toThrow("'version' length");
  });

  it("stores null for empty authors (convention-over-config: validated at connection time)", () => {
    const bot = new BotInfo(NAME, VERSION, [], null, null, [], [], null, null, null);
    expect(bot.authors).toBeNull();
  });

  it("throws if authors exceeds max count", () => {
    const tooMany = Array.from({ length: 21 }, (_, i) => `Author${i}`);
    expect(() => new BotInfo(NAME, VERSION, tooMany, null, null, [], [], null, null, null)).toThrow("'authors'");
  });

  it("throws if author name exceeds max length", () => {
    const longAuthor = "a".repeat(51);
    expect(() => new BotInfo(NAME, VERSION, [longAuthor], null, null, [], [], null, null, null)).toThrow("'author' length");
  });

  it("throws if description exceeds max length", () => {
    const longDesc = "a".repeat(251);
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, longDesc, null, [], [], null, null, null)).toThrow("'description' length");
  });

  it("throws if homepage exceeds max length", () => {
    const longUrl = "a".repeat(151);
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, longUrl, [], [], null, null, null)).toThrow("'homepage' length");
  });

  it("throws if country codes exceed max count", () => {
    const codes = ["US","GB","DE","FR","JP","AU","CA","BR","IN","CN","RU","MX","KR","IT","ES","PL","SE","NO","DK","NL","CH"];
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, null, codes, [], null, null, null)).toThrow("'countryCodes'");
  });

  it("silently drops invalid country codes", () => {
    const bot = new BotInfo(NAME, VERSION, AUTHORS, null, null, ["XX", "US"], [], null, null, null);
    expect(bot.countryCodes).toEqual(["US"]);
  });

  it("normalizes country codes to uppercase", () => {
    const bot = new BotInfo(NAME, VERSION, AUTHORS, null, null, ["us", "gb"], [], null, null, null);
    expect(bot.countryCodes).toEqual(["US", "GB"]);
  });

  it("throws if game types exceed max count", () => {
    const types = ["a","b","c","d","e","f","g","h","i","j","k"];
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, null, [], types, null, null, null)).toThrow("'gameTypes'");
  });

  it("throws if game type exceeds max length", () => {
    const longType = "a".repeat(21);
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, null, [], [longType], null, null, null)).toThrow("'gameTypes' length");
  });

  it("throws if platform exceeds max length", () => {
    const longPlatform = "a".repeat(31);
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, null, [], [], longPlatform, null, null)).toThrow("'platform' length");
  });

  it("throws if programmingLang exceeds max length", () => {
    const longLang = "a".repeat(31);
    expect(() => new BotInfo(NAME, VERSION, AUTHORS, null, null, [], [], null, longLang, null)).toThrow("'programmingLang' length");
  });

  it("trims description, homepage, platform, programmingLang", () => {
    const bot = new BotInfo(NAME, VERSION, AUTHORS, "  desc  ", "  http://x.com  ", [], [], "  Node  ", "  TS  ", null);
    expect(bot.description).toBe("desc");
    expect(bot.homepage).toBe("http://x.com");
    expect(bot.platform).toBe("Node");
    expect(bot.programmingLang).toBe("TS");
  });

  it("returns null for blank optional fields", () => {
    const bot = new BotInfo(NAME, VERSION, AUTHORS, "  ", "  ", [], [], "  ", "  ", null);
    expect(bot.description).toBeNull();
    expect(bot.homepage).toBeNull();
    expect(bot.platform).toBeNull();
    expect(bot.programmingLang).toBeNull();
  });
});
});
